src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -3757,45 +3757,78 @@ private void LearnFromEqualsMethod(MethodSymbol method, BoundCall node, TypeWith
                 {
-                    if (member.OriginalDefinition.Equals(wellKnownMethod))
+                    return true;
+                }
+
+                // check whether 'method', when called on this receiver, is an implementation of 'constructedMethod'.
+                for (var baseType = receiverType; baseType is object && method is object; baseType = baseType.BaseTypeNoUseSiteDiagnostics)
+                {
+                    var implementationMethod = baseType.FindImplementationForInterfaceMember(constructedMethod);
+                    if (implementationMethod is null)
                     {
-                        constructedMethod = member;
-                        break;
+                        // we know no base type will implement this interface member either
+                        return false;
                     }
-                }
 
-                Debug.Assert(constructedMethod != null, "the original definition is present but the constructed method isn't present");
+                    if (implementationMethod.ContainingType.IsInterface)
+                    {
+                        // this doesn't make sense because an interface can only explicitly implement a method from its base interface.
+                        return false;
+                    }
 
-                // FindImplementationForInterfaceMember doesn't check if this method is itself the interface method we're looking for
-                if (constructedMethod.Equals(method))
-                {
-                    return true;
+                    // could be calling an override of a method that implements the interface method
+                    for (var overriddenMethod = method; overriddenMethod is object; overriddenMethod = overriddenMethod.OverriddenMethod)
+                    {
+                        if (overriddenMethod.Equals(implementationMethod))
+                        {
+                            return true;
+                        }
+                    }
+
+                    // the Equals method being called isn't the method that implements the interface method in this type.
+                    // it could be a method that implements the interface on a base type, so check again with the base type.
+                    // if the method is directly contained in this type, we should skip up to its overridden method.
+
+                    // e.g. in this hierarchy:
+                    // class A -> B -> C -> D
+                    // method virtual B.Equals -> override D.Equals
+                    //
+                    // we would potentially check:
+                    // 1. D.Equals when called on D, then B.Equals when called on D
+                    // 2. B.Equals when called on C
+                    // 3. B.Equals when called on B
+                    // 4. give up when checking A, since B.Equals is not overriding anything in A
+
+                    if (baseType.Equals(method.ContainingType))
```

Comment:
>if (baseType.Equals(method.ContainingType)) [](start = 20, length = 43)

I think we should be able to do better here and, in some cases jump over multiple base types and overridden methods. Basically, the next base type on which we should call ```FindImplementationForInterfaceMember``` should be above the ```implementationMethod.ContainingType```. For any type in between, ```FindImplementationForInterfaceMember``` is going to return the same ```implementationMethod```, but we already checked it. Consider advancing ```baseType``` here in a loop until we get to ```implementationMethod.ContainingType```. While doing so, also advance ```method``` as we pass through the bases.

Reply:
Consider coming up with a test scenario and walking through this optimization under debugger to observe its effect.

---
In reply to: [390475202](https://github.com/dotnet/roslyn/pull/41918#discussion_r390475202) [](ancestors = 390475202)

Reply:
I think I understand now. Thanks for the explanation.